src/hotspot/share/gc/shenandoah/shenandoahStackWatermark.cpp

```diff
@@ -73,11 +74,11 @@ OopClosure* ShenandoahStackWatermark::closure_from_context(void* context) {
     assert(Thread::current()->is_Worker_thread(), "Unexpected thread passing in context: " PTR_FORMAT, p2i(context));
     return reinterpret_cast<OopClosure*>(context);
   } else {
-    if (_heap->is_concurrent_mark_in_progress()) {
-      return &_keep_alive_cl;
-    } else if (_heap->is_concurrent_weak_root_in_progress()) {
+    if (_heap->is_concurrent_weak_root_in_progress()) {
       assert(_heap->is_evacuation_in_progress(), "Nothing to evacuate");
       return &_evac_update_oop_cl;
+    } else if (_heap->is_concurrent_mark_in_progress()) {
+      return &_keep_alive_cl;
```

Comment:
Swapping conditions, again. Can concurrent mark overlap with concurrent roots, now, and this is why this is relevant?

Reply:
See https://github.com/openjdk/jdk/pull/20395#discussion_r1720473711